vtex-apps/return-app
# V3 create endpoint

This PR brings the API to create a return via HTTP request. It reuses the logic create for the GraphQL resolver. The flow was moved to a function `createReturnRequestService`. It also fixes the issue of requests without the cookie session getting the wrong user email saved as customer data. The logic for this can be reviewed here: `node/utils/getCostumerEmail.ts`.

### How to test it

Make a POST request to the following URL, passing either `APPKEY` and `APPTOKEN` or `VtexIdclientAutCookie` (the only difference for each case is how the `submittedBy` field is set - for the former it saves the appkey. For the latter it saves the user mail associated to the cookie). To see the request created, you can visit the admin or fetch the document via POSTMAN.

```
https://rmav3endpointsetup--powerplanet.myvtex.com/_v/return-request
```

Two ways for testing it:

#### The brave way

There are several error handlers in the flow. By calling it without any body, the errors should help you to build the body. It's longer for testing, but it helps testing the UX for the API too 😁

#### The fast way 

Pass the following body to the POST request. Only the first request should work. The following ones should fail since there won't be any product available to be returned in this order. We can delete the request hitting the endpoint:

```
DELETE http://powerplanet.vtexcommercestable.com.br/api/dataentities/vtex_return_app_returnRequest/documents/:documentId
```

```
{
    "items": [{
        "orderItemIndex": 0,
        "quantity": 1,
        "returnReason": {
            "reason": "Wrong type"
        },
        "condition": "newWithBox"
    }],
    "orderId": "1240221188059-01",
    "refundPaymentData": {
        "refundPaymentMethod": "sameAsPurchase"
    },
    "pickupReturnData": {
        "addressType": "CUSTOMER_ADDRESS",
        "address": "Rua Haddock Lobo",
        "city": "São Paulo",
        "state": "SP",
        "country": "Br",
        "zipCode": "01403003"
    },
    "customerProfileData": {
        "name": "Filadelfo Braz",
        "email": "filadelfo.braz+test@gmail.com",
        "phoneNumber": "123432122"
    },
    "userComment": "This is a test from API",
    "locale": "pt-PT"
}
```

### Body request doc:

```
orderId string required
items array of objects required
  Object:
    // Index of the item in the Order object form the OMS
    orderItemIndex integer required
    quantity integer required
    condition enum required
      Enum:
        newWithBox
        newWithoutBox
        usedWithBox
        usedWithoutBox
    returnReason object required
      Object:
        // If the store has custom reasons, this has to match one them
        reason string required
        otherReason string
customerProfileData object required
  Object:
    name string required
    email string required
    phoneNumber string required
pickupReturnData object required
  Object:
    addressId string required
    address string required
    city string required
    state string required
    country string required
    zipCode string required
    addressType enum required
      Enum:
        PICKUP_POINT
        CUSTOMER_ADDRESS
refundPaymentData object required
    Object:
      refundPaymentMethod enum required
        Enum:
          bank
          card
          giftCard
          sameAsPurchase
      iban string required when refundPaymentMethod is bank
      accountHolderName string required when refundPaymentMethod is bank
userComment string
locale string required
```
